pkg/kubelet/kubelet_node_status.go

```diff
@@ -31,24 +32,188 @@ import (
+	// updateRuntimeUpFunc is kubelet updateRuntimeUp function.
+	updateRuntimeUpFunc func()
+	// updatePodCIDRFunc is kubelet updatePodCIDR function.
+	updatePodCIDRFunc func(string) (bool, error)
+	// setLastObservedNodeAddressesFunc is kubelet setLastObservedNodeAddresses function.
+	setLastObservedNodeAddressesFunc func(addresses []v1.NodeAddress)
+	// keepTerminatedPodVolumes is kubelet keepTerminatedPodVolumes.
+	keepTerminatedPodVolumes bool
+}
+
+// NewNodeStatusManager build nodeStatus Manager.
+// TODO (follow up work)move it and related tests to kubelet/nodestatus package.
+func NewNodeStatusManager(
+	hostname string,
+	nodeName types.NodeName,
+	clock clock.Clock,
+	nodeLabels map[string]string,
+	registerNode bool,
+	kubeClient clientset.Interface,
+	heartbeatClient clientset.Interface,
+	providerID string,
+	externalCloudProvider bool,
+	containerManager cm.ContainerManager,
+	volumeManager volumemanager.VolumeManager,
+	cloud cloudprovider.Interface,
+	registerSchedulable bool,
+	registerWithTaints []api.Taint,
+	enableControllerAttachDetach bool,
+	onRepeatedHeartbeatFailure func(),
+	nodeStatusUpdateFrequency time.Duration,
+	nodeStatusReportFrequency time.Duration,
+	keepTerminatedPodVolumes bool,
+	getNodeFunc func() (*v1.Node, error),
+	updateRuntimeUpFunc func(),
+	updatePodCIDRFunc func(string) (bool, error),
+	setLastObservedNodeAddressesFunc func(addresses []v1.NodeAddress),
+	nodeStatusFuncs []func(*v1.Node) error,
+) *NodeStatusManager {
+	return &NodeStatusManager{
+		hostname:                         hostname,
+		nodeName:                         nodeName,
+		nodeLabels:                       nodeLabels,
+		clock:                            clock,
+		registerNode:                     registerNode,
+		kubeClient:                       kubeClient,
+		heartbeatClient:                  heartbeatClient,
+		providerID:                       providerID,
+		externalCloudProvider:            externalCloudProvider,
+		containerManager:                 containerManager,
+		volumeManager:                    volumeManager,
+		cloud:                            cloud,
+		registerSchedulable:              registerSchedulable,
+		registerWithTaints:               registerWithTaints,
+		enableControllerAttachDetach:     enableControllerAttachDetach,
+		onRepeatedHeartbeatFailure:       onRepeatedHeartbeatFailure,
+		nodeStatusUpdateFrequency:        nodeStatusUpdateFrequency,
+		nodeStatusReportFrequency:        nodeStatusReportFrequency,
+		keepTerminatedPodVolumes:         keepTerminatedPodVolumes,
+		getNodeFunc:                      getNodeFunc,
+		updateRuntimeUpFunc:              updateRuntimeUpFunc,
+		updatePodCIDRFunc:                updatePodCIDRFunc,
+		setLastObservedNodeAddressesFunc: setLastObservedNodeAddressesFunc,
+		nodeStatusFuncs:                  nodeStatusFuncs,
+	}
+}
+
+// Run sync node status.
+func (s *NodeStatusManager) Run(stopCh <-chan struct{}) {
+	go s.fastStatusUpdateOnce()
```

Comment:
if possible, this should also respect the stopCh.

Reply:
I haven't come up with a method which return once node's podCIDR updated in wait forever condition. If fastStatusUpdateOnce logic can be changed, I prefer to do it in a follow up work(I'm not intend to break current functions in this pr).

Reply:
my suggestion would be to select over the sleep and the stopCh.  E.g.

```golang
for {
    select {
    case <-stopCh:
        return
    case <-time.After(100*time.Millisecond):
    }
    ...
}
```

But i'm also OK with this in a follow-up since the PR is large.

Reply:
Got it, thank you.